File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/pushTokenRegistration/PushTokenRegistrationHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.pushTokenRegistration

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.pushTokenRegistration.PushTokenRegistrationApi
import features.pushTokenRegistration.models.UserProfileModel

import features.pushTokenRegistration.serializePushTokenRegistrationState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class PushTokenRegistrationHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/register_push_token"

    override val eventChannel: String
        get() = "customer_sdk/events/register_push_token"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return registerPushToken(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun registerPushToken(call: MethodCall): String? {
            val result = PushTokenRegistrationApi().registerPushToken(
                androidDeviceToken = call.argument<String?>("androidDeviceToken"),
                iosDeviceToken = call.argument<String?>("iosDeviceToken")
            ).toJsonResult<UserProfileModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        PushTokenRegistrationApi().pushTokenRegistrationState.serializePushTokenRegistrationState().observe { result ->
            pushSerializedResultToEventSink<PushTokenRegistrationApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/pushTokenRegistration/PushTokenRegistrationHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class PushTokenRegistrationHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/register_push_token"

    var eventChannel: String = "customer_sdk/events/register_push_token"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await PushTokenRegistrationHandler.registerPushToken(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func registerPushToken(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        let androidDeviceToken = myArgs["androidDeviceToken"] as? String? ?? nil
                let iosDeviceToken = myArgs["iosDeviceToken"] as? String? ?? nil

         let result = try await PushTokenRegistrationApi().registerPushToken(
                androidDeviceToken: androidDeviceToken,
                iosDeviceToken: iosDeviceToken
            ).serializeNative()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        PushTokenRegistrationApi().pushTokenRegistrationState
            .serializePushTokenRegistrationState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
